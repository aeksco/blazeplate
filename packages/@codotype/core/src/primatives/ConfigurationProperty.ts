import { v4 as uuidv4 } from "uuid";
import {
    ConfigurationProperty,
    DropdownOption,
    OptionType,
    OptionValue,
    PropertyLayoutVariants,
} from "../configuration-option-types";
import { PropertyTransformation } from "../property-transformation";
import { PropertyValidation } from "../property-validation";
import { DataPreviewLayoutVariant, DataPreview } from "../data-preview";
import { Content } from "../content";

// // // //

interface ConfigurationPropertyBuilderParams {
    id?: string;
    identifier: string;
    content: {
        label: string;
        description?: string;
        documentation?: string;
        icon?: string;
    };
    type: OptionType;
    dropdownOptions?: DropdownOption[];
    defaultValue?: OptionValue;
    required?: boolean;
    enabledByDefault?: boolean;
    allowDisable?: boolean;
    layoutVariant?: PropertyLayoutVariants;
    properties?: ConfigurationProperty[];
    transformations?: PropertyTransformation[];
    validations?: PropertyValidation[];
    dataPreview?: DataPreview;
}

export class ConfigurationPropertyBuilder implements ConfigurationProperty {
    id: string = uuidv4();
    identifier: string;
    content: Content = {
        label: "",
        description: "",
        documentation: "",
        icon: "",
    };
    type: OptionType;
    defaultValue: OptionValue = null;
    required: boolean = false;
    unique: boolean = false;
    enabledByDefault: boolean = true;
    allowDisable: boolean = false;
    layoutVariant: PropertyLayoutVariants;
    properties: ConfigurationProperty[] = [];
    dropdownOptions: DropdownOption[] = [];
    transformations: PropertyTransformation[] = [];
    validations: PropertyValidation[] = [];
    dataPreview: DataPreview = {
        rules: [],
        variant: DataPreviewLayoutVariant.CODE_DARK,
    };

    constructor(params: ConfigurationPropertyBuilderParams) {
        this.id = params.id || this.id;
        this.identifier = params.identifier;

        // Sets this.content
        this.content = {
            label: params.content.label,
            description: params.content.description || this.content.description,
            documentation:
                params.content.documentation || this.content.documentation,
            icon: params.content.icon || this.content.icon,
        };

        this.type = params.type;
        this.defaultValue = params.defaultValue || this.defaultValue;
        this.required = params.required || this.required;
        this.enabledByDefault =
            params.enabledByDefault || this.enabledByDefault;
        this.allowDisable = params.allowDisable || this.allowDisable;
        this.layoutVariant =
            params.layoutVariant || PropertyLayoutVariants.CARD_COL_12;
        this.properties = params.properties || this.properties;
        this.dropdownOptions = params.dropdownOptions || this.dropdownOptions;
        this.validations = params.validations || this.validations;
        this.transformations = params.transformations || this.transformations;
        this.dataPreview = params.dataPreview || this.dataPreview;
    }
}
