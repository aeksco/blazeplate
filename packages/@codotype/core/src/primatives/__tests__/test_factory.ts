import { PropertyType } from "../../configuration-property";
import { ConfigurationPropertyBuilder } from "../ConfigurationProperty";

// Mocks UUID module
jest.mock("uuid", () => ({
    __esModule: true,
    v4: () => "1234-5678-1234-5678",
}));

// // // //

describe("factory", () => {
    test("works", () => {
        const configurationGroupProperty = new ConfigurationPropertyBuilder({
            identifier: "meh_value",
            type: PropertyType.BOOLEAN,
            content: {
                label: "meh",
            },
        });

        expect(configurationGroupProperty).toMatchSnapshot();
    });
});
