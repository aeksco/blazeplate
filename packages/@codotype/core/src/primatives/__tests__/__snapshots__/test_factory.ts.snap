// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`factory works 1`] = `
ConfigurationGroupPropertyBuilder {
  "allowDisable": false,
  "content": Object {
    "description": "",
    "documentation": "",
    "icon": "",
    "label": "meh",
  },
  "dataPreview": Object {
    "rules": Array [],
    "variant": "CODE_DARK",
  },
  "defaultValue": null,
  "dropdownOptions": Array [],
  "enabledByDefault": true,
  "id": "1234-5678-1234-5678",
  "identifier": "meh_value",
  "layoutVariant": "CARD_COL_12",
  "properties": Array [],
  "required": false,
  "transformations": Array [],
  "type": "BOOLEAN",
  "unique": false,
  "validations": Array [],
}
`;
